presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java

```diff
@@ -0,0 +1,440 @@
+
+        TypeSignature expectedReturnTypeSignature = declaredSignature.getReturnType();
+        if (!matchAndBindArgument(expectedReturnTypeSignature, actualReturnType, boundVariablesBuilder)) {
+            return Optional.empty();
+        }
+
+        List<TypeSignature> expectedArgumentSignatures = declaredSignature.getArgumentTypes();
+        boolean variableArity = declaredSignature.isVariableArity();
+        if (!matchArguments(expectedArgumentSignatures, actualArgumentTypes, variableArity, boundVariablesBuilder)) {
+            return Optional.empty();
+        }
+
+        calculateVariableValuesForLongConstraints(boundVariablesBuilder);
+
+        BoundVariables boundVariables = boundVariablesBuilder.build();
+        checkAllTypeVariablesAreBound(boundVariables);
+        return Optional.of(boundVariables);
+    }
+
+    private boolean matchArguments(
+            List<TypeSignature> expectedArgumentSignatures,
+            List<? extends Type> actualArgumentTypes,
+            boolean variableArity,
+            BoundVariables.Builder boundVariables)
+    {
+        if (variableArity) {
+            if (actualArgumentTypes.size() < expectedArgumentSignatures.size() - 1) {
+                return false;
+            }
+        }
+        else {
+            if (expectedArgumentSignatures.size() != actualArgumentTypes.size()) {
+                return false;
+            }
+        }
+
+        // Bind the variable arity argument first, to make sure it's bound to the common super type
+        if (variableArity && actualArgumentTypes.size() >= expectedArgumentSignatures.size()) {
+            List<? extends Type> variableArityParameterTypes = actualArgumentTypes.subList(
+                    expectedArgumentSignatures.size() - 1, actualArgumentTypes.size()
+            );
+            Optional<Type> variableArityParameterActualType = typeManager.getCommonSuperType(variableArityParameterTypes);
+            if (!variableArityParameterActualType.isPresent()) {
+                return false;
+            }
+            TypeSignature expectedVariableArityParameterSignature = expectedArgumentSignatures.get(expectedArgumentSignatures.size() - 1);
+            if (!matchAndBindArgument(expectedVariableArityParameterSignature, variableArityParameterActualType.get(), boundVariables)) {
+                return false;
+            }
+        }
+
+        for (int i = 0; i < actualArgumentTypes.size(); i++) {
+            // Get the current argument signature, or the last one, if this is a varargs function
+            TypeSignature expectedArgumentSignature = expectedArgumentSignatures.get(Math.min(i, expectedArgumentSignatures.size() - 1));
+            Type actualArgumentType = actualArgumentTypes.get(i);
+            if (!matchAndBindArgument(expectedArgumentSignature, actualArgumentType, boundVariables)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private boolean matchAndBindArgument(
```

Comment:
Rename to `bind`. Also, name the arguments `expected`, `actual` and `variableBinder`